### PR TITLE
End to end test workflow

### PR DESCRIPTION
**Pull request by Brainicism**
EDIT by @grqz: pinned info:
examples:
* RH's not invoked https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10869148543/job/30159954418
* Passing: https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10870654853/job/30163532307
* HTTP Server not transpiling: https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10869223803/job/30160127984
* HTTP Server terminates midway after being pinged: https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10869309089/job/30160329617
>[!NOTE]
> some of them may not be up-to-date

### CODE REVIEW

**Comment by grqz** [comment]
let's move the `lint and fmt check` here

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        run: curl -L https://github.com/yt-dlp/yt-dlp-nightly-builds/releases/latest/download/yt-dlp -o yt-dlp && chmod +x ./yt-dlp
+      - name: Install and build server dependencies
+        run: cd server/ && yarn install --frozen-lockfile && npx tsc
+      - name: Install plugin
+        run: |
+          mkdir ~/yt-dlp-plugins && curl -L "https://github.com/coletdjnz/yt-dlp-get-pot/releases/download/v0.1.1/yt-dlp-get-pot.zip" -o ~/yt-dlp-plugins/yt-dlp-get-pot.zip
+          cp -aT plugin/ ~/yt-dlp-plugins/bgutil-ytdlp-pot-provider/
+      - name: Test script method
+        run: |
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" 2>&1)
```
Comment (grqz) [inline .github/workflows/test.yml:26]: consider using default path

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+          echo $script_response
+          if [[ "$script_response" != *"BgUtilScriptPot: _get_pot_via_script response"* ]]; then
+            echo "BgUtilScriptPot was not invoked"
+            exit 1
+          fi
+      - name: Test server method
+        run: |
+          cd server/
+          node build/main.js&
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_baseurl=http://127.0.0.1:4416 2>&1")
```
Comment (grqz) [inline .github/workflows/test.yml:36]: the same here

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+          if [[ "$script_response" != *"BgUtilScriptPot: _get_pot_via_script response"* ]]; then
+            echo "BgUtilScriptPot was not invoked"
+            exit 1
+          fi
+      - name: Test server method
+        run: |
+          cd server/
+          node build/main.js&
+          cd ..
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_baseurl=http://127.0.0.1:4416 2>&1")
```
Comment (grqz) [inline .github/workflows/test.yml:37]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_baseurl=http://127.0.0.1:4416" 2>&1)
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -34,6 +34,10 @@ jobs:
           cd server/
           node build/main.js&
           cd ..
+          until curl --silent --fail http://127.0.0.1:4416/ping; do
+            echo "Waiting for server to be up..."
+            sleep 5
```
Comment (grqz) [inline .github/workflows/test.yml:39]: maybe ~1sec is enough

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          node build/main.js&
+          cd ..
+          until curl --silent --fail http://127.0.0.1:4416/ping; do
+            echo "Waiting for server to be up..."
+            sleep 5
+          done
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_baseurl=http://127.0.0.1:4416" 2>&1)
+          echo $script_response
+          if [[ "$script_response" != *"BgUtilHTTPPot: Generating POT via HTTP server"* ]]; then
+            echo "BgUtilHTTPPot was not invoked"
```
Comment (grqz) [inline .github/workflows/test.yml:44]: consider using github format

Comment (grqz) [reply .github/workflows/test.yml:44]: see: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-error-message

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            echo "BgUtilScriptPot was not invoked"
+            exit 1
+          fi
+      - name: Test server method
+        run: |
+          cd server/
+          node build/main.js&
+          cd ..
+          until curl --silent --fail http://127.0.0.1:4416/ping; do
+            echo "Waiting for server to be up..."
```
Comment (grqz) [inline .github/workflows/test.yml:38]: this should be printed only once imo

**Comment by grqz** [comment]
are we still going to run the `build docker image` on prs? test.yml can now test whether the script is transpiling

**Comment by Brainicism** [comment]
> are we still going to run the `build docker image` on prs? test.yml can now test whether the script is transpiling. let's make `test` pure test and `create release` pure `create release`

Sounds like a good idea to me. Only difference I can see if there's ever any discrepancy between the Docker base image's environment, and the github runner image that causes build to fail in one but not the other. Seems unlikely though

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+          timeout 7.5 bash -c "ping 0.5" || \
+            (echo "::error title=Timeout reached,file=server/src/main.ts::\
+          Timeout reached before the server is up." && exit 1)
+
+          echo "::group::Response from HTTP server"
+          echo $RESP | jq
+          echo "::endgroup::"
+
+          cd ..
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:88]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate BaW_jenozKc 2>&1) || \
```

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Install and build server dependencies
+        run: cd server/ && yarn install --frozen-lockfile && npx tsc
+      - name: Install plugin
+        run: |
+          mkdir ~/yt-dlp-plugins && curl -L "https://github.com/coletdjnz/yt-dlp-get-pot/releases/download/v0.1.1/yt-dlp-get-pot.zip" -o ~/yt-dlp-plugins/yt-dlp-get-pot.zip
+          cp -aT plugin/ ~/yt-dlp-plugins/bgutil-ytdlp-pot-provider/
+      - name: Test script method
+        shell: bash
+        run: |
+          script_response=$(./yt-dlp -vF --print-traffic BaW_jenozKc --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:50]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate BaW_jenozKc --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" 2>&1) || \
```

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+          printf "%s\n" "$script_response"
+          echo "::endgroup::"
+          if [[ "$script_response" != *"BgUtilScriptPot: Generating POT via script: "* ]]; then
+            echo "::error title=BgUtilScriptPot was not invoked,file=plugin/yt_dlp_plugins/extractor/getpot_bgutil_script.py::\
+          BgUtilScriptPot was not invoked"
+            exit 1
+          fi
+      - name: Test server method
+        timeout-minutes: 3
+        if: always()
```
Comment (Brainicism) [inline .github/workflows/test.yml:64]: why is this if always?

Comment (grqz) [reply .github/workflows/test.yml:64]: the same as those `Lint and format check`, script and server are independent.
maybe break it if `Install and build server dependencies` fails

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+          timeout 7.5 bash -c "ping 0.5" || \
+            (echo "::error title=Timeout reached,file=server/src/main.ts::\
+          Timeout reached before the server is up." && exit 1)
+
+          echo "::group::Response from HTTP server"
+          echo $RESP | jq
+          echo "::endgroup::"
+
+          cd ..
+          script_response=$(./yt-dlp -vF --print-traffic --no-simulate BaW_jenozKc 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:88]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" dQw4w9WgXcQ 2>&1) || \
```

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Install and build server dependencies
+        run: cd server/ && yarn install --frozen-lockfile && npx tsc
+      - name: Install plugin
+        run: |
+          mkdir ~/yt-dlp-plugins && curl -L "https://github.com/coletdjnz/yt-dlp-get-pot/releases/download/v0.1.1/yt-dlp-get-pot.zip" -o ~/yt-dlp-plugins/yt-dlp-get-pot.zip
+          cp -aT plugin/ ~/yt-dlp-plugins/bgutil-ytdlp-pot-provider/
+      - name: Test script method
+        shell: bash
+        run: |
+          script_response=$(./yt-dlp -vF --print-traffic --no-simulate BaW_jenozKc --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:50]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" dQw4w9WgXcQ 2>&1) || \
```

File: .github/workflows/test.yml [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Install and build server dependencies
+        run: cd server/ && yarn install --frozen-lockfile && npx tsc
+      - name: Install plugin
+        run: |
+          mkdir ~/yt-dlp-plugins && curl -L "https://github.com/coletdjnz/yt-dlp-get-pot/releases/download/v0.1.1/yt-dlp-get-pot.zip" -o ~/yt-dlp-plugins/yt-dlp-get-pot.zip
+          cp -aT plugin/ ~/yt-dlp-plugins/bgutil-ytdlp-pot-provider/
+      - name: Test script method
+        shell: bash
+        run: |
+          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" dQw4w9WgXcQ 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:50]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" --exec "rm -f" --extractor-args "youtube:getpot_bgutil_script=server/build/generate_once.js" dQw4w9WgXcQ 2>&1) || \
```

File: .github/workflows/test.yml [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+          PING_RESP=$(timeout 7.5 bash -c ping) || \
+            (echo "::error title=Timeout reached,file=server/src/main.ts::\
+          Timeout reached before the server is up." && exit 1)
+
+          echo "::group::Response from HTTP server"
+          echo "$PING_RESP" | jq
+          echo "::endgroup::"
+
+          cd ..
+          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" dQw4w9WgXcQ 2>&1) || \
```
Comment (grqz) [inline .github/workflows/test.yml:88]:
```suggestion
          script_response=$(./yt-dlp -vF --print-traffic --no-simulate -f "b*" -S "filesize:10M" --exec "rm -f" dQw4w9WgXcQ 2>&1) || \
```

File: .github/workflows/test.yml [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+          context: server/
+          file: server/Dockerfile
+          push: false
+          tags: brainicism/bgutil-ytdlp-pot-provider:ci
+      - name: Test server method
+        timeout-minutes: 3
+        shell: bash
+        run: |
+          docker run --name bgutil-provider -d -p 4416:4416 brainicism/bgutil-ytdlp-pot-provider:ci
+          NODE_PID=$!
```
Comment (grqz) [inline .github/workflows/test.yml:86]: NODE_PID is not used here, should we run docker in parallel and use docker stop instead?

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+          context: server/
+          file: server/Dockerfile
+          push: false
+          tags: brainicism/bgutil-ytdlp-pot-provider:ci
+      - name: Test server method
+        timeout-minutes: 3
+        shell: bash
+        run: |
+          docker run --name bgutil-provider -d -p 4416:4416 brainicism/bgutil-ytdlp-pot-provider:ci
+          echo "Waiting for server to be up..."
```
Comment (grqz) [inline .github/workflows/test.yml:86]: let's move this `echo` into func `ping` for better code readability

File: .github/workflows/test.yml [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+      - name: Install plugin
+        run: |
+          mkdir ~/yt-dlp-plugins && curl -L "https://github.com/coletdjnz/yt-dlp-get-pot/releases/download/v0.1.1/yt-dlp-get-pot.zip" -o ~/yt-dlp-plugins/yt-dlp-get-pot.zip
+          cp -aT plugin/ ~/yt-dlp-plugins/bgutil-ytdlp-pot-provider/
+      - name: Build Docker image
+        uses: docker/build-push-action@v6
+        with:
+          context: server/
+          file: server/Dockerfile
+          push: false
```
Comment (grqz) [inline .github/workflows/test.yml:79]: `push: false` is the default val and we can omit it, does passing [`load: true`](https://docs.docker.com/engine/reference/commandline/buildx_build/#push) help here?

File: .github/workflows/test.yml [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+          tags: brainicism/bgutil-ytdlp-pot-provider:ci
```
Comment (grqz) [inline .github/workflows/test.yml:80]: this tag is the same as the release one. consider removing/changing either of them to clarify

**Comment by grqz** [comment]
why is https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10875000151/job/30172911791 failing?

**Comment by Brainicism** [comment]
> why is https://github.com/Brainicism/bgutil-ytdlp-pot-provider/actions/runs/10875000151/job/30172911791 failing?

not sure..

**Comment by grqz** [comment]
i see. it seems to be because we're giving the response of function ping to jq. we should still keep the echo command out. probably before the `timeout`

**Review by grqz — APPROVED** [review]
LGTM

**Pull request merged**